File: FirelinkER/tests/FirelinkERMaps/test_er_msb.cpp
// test_er_msb.cpp — comprehensive doctest suite for Elden Ring MSB read/write.
//
// Uses the bundled Stormveil Castle MSB (m10_00_00_00.msb, DCX removed) to verify:
//   1.  MSB loads without error.
//   2.  All param counts are positive.
//   3.  Known subtype distributions are non-empty.
//   4.  Names, types, and cross-references resolve correctly.
//   5.  String conversion operator works on entries.
//   6.  Draw-parent cross-references can be mutated.
//   7.  Write -> re-read produces identical param counts and entry data.
//   8.  Double write produces byte-identical files (stable serialization).

#include <doctest/doctest.h>

#include <FirelinkCore/BinaryReadWrite.h>
#include <FirelinkERMaps/MapStudio/EntryParam.h>
#include <FirelinkERMaps/MapStudio/MSB.h>
#include <FirelinkERMaps/MapStudio/Model.h>
#include <FirelinkERMaps/MapStudio/Event.h>
#include <FirelinkERMaps/MapStudio/Region.h>
#include <FirelinkERMaps/MapStudio/Part.h>

#include <chrono>
#include <filesystem>
#include <fstream>
#include <memory>
#include <string>
#include <vector>

using namespace std;
using namespace Firelink;
using namespace FirelinkER::Maps::MapStudio;

// ============================================================================
// Helpers
// ============================================================================

namespace
{
    /// Resolve the bundled MSB resource path.
    filesystem::path GetMSBPath()
    {
        filesystem::path p = filesystem::path(TEST_RESOURCES_DIR) / "m10_00_00_00.msb";
        REQUIRE_MESSAGE(exists(p), "MSB resource not found: ", p.string());
        return p;
    }

    /// Lazily load and cache the original MSB (shared across read-only tests).
    MSB& GetOriginalMSB()
    {
        static unique_ptr<MSB> msb = [] {
            const auto start = chrono::high_resolution_clock::now();
            auto ptr = MSB::NewFromFilePath(GetMSBPath());
            const auto elapsed = chrono::duration_cast<chrono::milliseconds>(
                chrono::high_resolution_clock::now() - start);
            MESSAGE("MSB loaded in " << elapsed.count() << " ms");
            return ptr;
        }();
        return *msb;
    }

    /// Read a whole file into a byte vector.
    vector<uint8_t> ReadWholeFile(const filesystem::path& path)
    {
        ifstream f(path, ios::binary | ios::ate);
        REQUIRE_MESSAGE(f.good(), "Cannot open file: ", path.string());
        const auto size = static_cast<size_t>(f.tellg());
        f.seekg(0);
        vector<uint8_t> buf(size);
        f.read(reinterpret_cast<char*>(buf.data()), static_cast<streamsize>(size));
        return buf;
    }

    /// Create (and cache) a temp output directory for round-trip files.
    const filesystem::path& GetOutDir()
    {
        static const filesystem::path dir = [] {
            auto d = filesystem::temp_directory_path() / "firelink_msb_test";
            create_directories(d);
            return d;
        }();
        return dir;
    }
} // namespace

// ============================================================================
// Basic loading
// ============================================================================

TEST_CASE("MSB: loads Stormveil Castle without error")
{
    CHECK_NOTHROW(GetOriginalMSB());
}

// ============================================================================
// Param counts
// ============================================================================

TEST_CASE("MSB: all param counts are positive")
{
    auto& msb = GetOriginalMSB();
    CHECK(msb.GetModelParam().GetSize()  > 0);
    CHECK(msb.GetEventParam().GetSize()  > 0);
    CHECK(msb.GetRegionParam().GetSize() > 0);
    CHECK(msb.GetRouteParam().GetSize()  > 0);
    CHECK(msb.GetPartParam().GetSize()   > 0);
}

// ============================================================================
// Subtype distributions
// ============================================================================

TEST_CASE("MSB: model subtype distribution is non-empty")
{
    auto& msb = GetOriginalMSB();
    auto all = msb.GetModelParam().GetAllEntries();

    int mapPiece = 0, character = 0, collision = 0, asset = 0;
    for (const auto* m : all)
    {
        switch (m->GetType())
        {
            case ModelType::MapPiece:  ++mapPiece;  break;
            case ModelType::Character: ++character;  break;
            case ModelType::Collision: ++collision;  break;
            case ModelType::Asset:     ++asset;      break;
            default: break;
        }
    }
    CHECK(mapPiece  > 0);
    CHECK(character > 0);
    CHECK(collision > 0);
    CHECK(asset     > 0);
}

TEST_CASE("MSB: part subtype distribution is non-empty")
{
    auto& msb = GetOriginalMSB();
    auto all = msb.GetPartParam().GetAllEntries();

    int mapPiece = 0, character = 0, collision = 0, asset = 0;
    for (const auto* p : all)
    {
        switch (p->GetType())
        {
            case PartType::MapPiece:  ++mapPiece;  break;
            case PartType::Character: ++character;  break;
            case PartType::Collision: ++collision;  break;
            case PartType::Asset:     ++asset;      break;
            default: break;
        }
    }
    CHECK(mapPiece  > 0);
    CHECK(character > 0);
    CHECK(collision > 0);
    CHECK(asset     > 0);
}

TEST_CASE("MSB: region count > 10 for Stormveil")
{
    auto& msb = GetOriginalMSB();
    CHECK(msb.GetRegionParam().GetAllEntries().size() > 10);
}

TEST_CASE("MSB: at least one Treasure event")
{
    auto& msb = GetOriginalMSB();
    auto all = msb.GetEventParam().GetAllEntries();
    int treasures = 0;
    for (const auto* e : all)
        if (e->GetType() == EventType::Treasure) ++treasures;
    CHECK(treasures > 0);
}

// ============================================================================
// Entry names and cross-references
// ============================================================================

TEST_CASE("MSB: all entry names are non-empty except Routes")
{
    auto& msb = GetOriginalMSB();

    for (const auto* m : msb.GetModelParam().GetAllEntries())
        CHECK_FALSE(m->GetNameUTF8().empty());
    for (const auto* e : msb.GetEventParam().GetAllEntries())
        CHECK_FALSE(e->GetNameUTF8().empty());
    for (const auto* r : msb.GetRegionParam().GetAllEntries())
        CHECK_FALSE(r->GetNameUTF8().empty());
    for (const auto* p : msb.GetPartParam().GetAllEntries())
        CHECK_FALSE(p->GetNameUTF8().empty());
}

TEST_CASE("MSB: every part has a model reference")
{
    auto& msb = GetOriginalMSB();
    for (const auto* p : msb.GetPartParam().GetAllEntries())
        CHECK_MESSAGE(p->GetModel() != nullptr, "Part '", p->GetNameUTF8(), "' has no model");
}

TEST_CASE("MSB: first part's model exists in model list")
{
    auto& msb = GetOriginalMSB();
    auto allParts  = msb.GetPartParam().GetAllEntries();
    auto allModels = msb.GetModelParam().GetAllEntries();
    REQUIRE_FALSE(allParts.empty());
    REQUIRE(allParts[0]->GetModel() != nullptr);

    const string modelName = allParts[0]->GetModel()->GetNameUTF8();
    bool found = false;
    for (const auto* m : allModels)
        if (m->GetNameUTF8() == modelName) { found = true; break; }
    CHECK(found);
}

// ============================================================================
// String conversion operator (from old er_msb_test.cpp)
// ============================================================================

TEST_CASE("MSB: entry string conversion produces non-empty output")
{
    auto& msb = GetOriginalMSB();
    auto allParts = msb.GetPartParam().GetAllEntries();
    REQUIRE_FALSE(allParts.empty());

    // Part -> string
    CHECK_FALSE(string(*allParts[0]).empty());

    // Character -> string
    auto characters = msb.GetPartParam().GetSubtypeEntries<CharacterPart>();
    REQUIRE_FALSE(characters.empty());
    CHECK_FALSE(string(*characters[0]).empty());

    // Model -> string
    if (const auto* model = characters[0]->GetModel())
        CHECK_FALSE(string(*model).empty());
}

// ============================================================================
// Character part fields
// ============================================================================

TEST_CASE("MSB: character part fields are valid")
{
    auto& msb = GetOriginalMSB();
    auto characters = msb.GetPartParam().GetSubtypeEntries<CharacterPart>();
    REQUIRE_FALSE(characters.empty());

    const auto* chr = characters[0];

    SUBCASE("translate is finite")
    {
        auto [x, y, z] = chr->GetTranslate();
        CHECK(isfinite(x));
        CHECK(isfinite(y));
        CHECK(isfinite(z));
    }

    SUBCASE("scale is positive")
    {
        auto [sx, sy, sz] = chr->GetScale();
        CHECK(sx > 0);
        CHECK(sy > 0);
        CHECK(sz > 0);
    }
}

// ============================================================================
// Collision part fields
// ============================================================================

TEST_CASE("MSB: collision part fields are valid")
{
    auto& msb = GetOriginalMSB();
    auto collisions = msb.GetPartParam().GetSubtypeEntries<CollisionPart>();
    REQUIRE_FALSE(collisions.empty());

    const auto* col = collisions[0];
    const uint8_t hf = col->GetHitFilterId();
    CHECK(hf >= 8);
    CHECK(hf <= 29);
}

// ============================================================================
// Region shapes
// ============================================================================

TEST_CASE("MSB: all regions have a shape object")
{
    auto& msb = GetOriginalMSB();
    auto allRegions = msb.GetRegionParam().GetAllEntries();
    for (auto* r : allRegions)
        CHECK_MESSAGE(r->GetShape() != nullptr, "Region '", r->GetNameUTF8(), "' has no shape");
}

// ============================================================================
// Draw-parent mutation (from old er_msb_test.cpp)
// ============================================================================

TEST_CASE("MSB: draw parent can be cleared and reassigned")
{
    // Load a fresh copy so mutations don't pollute other tests.
    auto msb = MSB::NewFromFilePath(GetMSBPath());
    auto allParts   = msb->GetPartParam().GetAllEntries();
    auto characters = msb->GetPartParam().GetSubtypeEntries<CharacterPart>();
    REQUIRE_FALSE(allParts.empty());
    REQUIRE_FALSE(characters.empty());

    auto* chr = characters[0];

    SUBCASE("clear draw parent")
    {
        chr->SetDrawParent(nullptr);
        CHECK(chr->GetDrawParent() == nullptr);
    }

    SUBCASE("set draw parent to another part")
    {
        auto* firstPart = allParts[0];
        chr->SetDrawParent(firstPart);
        REQUIRE(chr->GetDrawParent() != nullptr);
        CHECK(chr->GetDrawParent()->GetNameUTF8() == firstPart->GetNameUTF8());
    }
}

// ============================================================================
// Round-trip: write -> re-read -> compare
// ============================================================================

TEST_CASE("MSB: round-trip preserves param counts")
{
    auto& original = GetOriginalMSB();
    const auto writePath = GetOutDir() / "m10_00_00_00_roundtrip.msb";
    original.WriteToFilePath(writePath);
    auto reloaded = MSB::NewFromFilePath(writePath);
    REQUIRE(reloaded != nullptr);

    CHECK(original.GetModelParam().GetSize()  == reloaded->GetModelParam().GetSize());
    CHECK(original.GetEventParam().GetSize()  == reloaded->GetEventParam().GetSize());
    CHECK(original.GetRegionParam().GetSize() == reloaded->GetRegionParam().GetSize());
    CHECK(original.GetRouteParam().GetSize()  == reloaded->GetRouteParam().GetSize());
    CHECK(original.GetPartParam().GetSize()   == reloaded->GetPartParam().GetSize());
}

TEST_CASE("MSB: round-trip preserves entry names")
{
    auto& original = GetOriginalMSB();
    const auto writePath = GetOutDir() / "m10_00_00_00_roundtrip.msb";
    original.WriteToFilePath(writePath);
    auto reloaded = MSB::NewFromFilePath(writePath);
    REQUIRE(reloaded != nullptr);

    SUBCASE("part names")
    {
        auto partsA = original.GetPartParam().GetAllEntries();
        auto partsB = reloaded->GetPartParam().GetAllEntries();
        REQUIRE(partsA.size() == partsB.size());
        for (size_t i = 0; i < partsA.size(); ++i)
            CHECK(partsA[i]->GetNameUTF8() == partsB[i]->GetNameUTF8());
    }

    SUBCASE("model names")
    {
        auto a = original.GetModelParam().GetAllEntries();
        auto b = reloaded->GetModelParam().GetAllEntries();
        REQUIRE(a.size() == b.size());
        for (size_t i = 0; i < a.size(); ++i)
            CHECK(a[i]->GetNameUTF8() == b[i]->GetNameUTF8());
    }

    SUBCASE("event names")
    {
        auto a = original.GetEventParam().GetAllEntries();
        auto b = reloaded->GetEventParam().GetAllEntries();
        REQUIRE(a.size() == b.size());
        for (size_t i = 0; i < a.size(); ++i)
            CHECK(a[i]->GetNameUTF8() == b[i]->GetNameUTF8());
    }

    SUBCASE("region names")
    {
        auto a = original.GetRegionParam().GetAllEntries();
        auto b = reloaded->GetRegionParam().GetAllEntries();
        REQUIRE(a.size() == b.size());
        for (size_t i = 0; i < a.size(); ++i)
            CHECK(a[i]->GetNameUTF8() == b[i]->GetNameUTF8());
    }
}

TEST_CASE("MSB: round-trip preserves character fields")
{
    auto& original = GetOriginalMSB();
    const auto writePath = GetOutDir() / "m10_00_00_00_roundtrip.msb";
    original.WriteToFilePath(writePath);
    auto reloaded = MSB::NewFromFilePath(writePath);
    REQUIRE(reloaded != nullptr);

    auto charsOrig     = original.GetPartParam().GetSubtypeEntries<CharacterPart>();
    auto charsReloaded = reloaded->GetPartParam().GetSubtypeEntries<CharacterPart>();
    REQUIRE(charsOrig.size() == charsReloaded.size());

    for (size_t i = 0; i < charsOrig.size(); ++i)
    {
        CHECK(charsOrig[i]->GetAiId()        == charsReloaded[i]->GetAiId());
        CHECK(charsOrig[i]->GetCharacterId() == charsReloaded[i]->GetCharacterId());
        CHECK(charsOrig[i]->GetTalkId()      == charsReloaded[i]->GetTalkId());
        CHECK(charsOrig[i]->GetEntityId()    == charsReloaded[i]->GetEntityId());

        // Translate.
        auto [ox, oy, oz] = charsOrig[i]->GetTranslate();
        auto [rx, ry, rz] = charsReloaded[i]->GetTranslate();
        CHECK(ox == rx);
        CHECK(oy == ry);
        CHECK(oz == rz);

        // Draw parent cross-reference.
        const auto* dpOrig     = charsOrig[i]->GetDrawParent();
        const auto* dpReloaded = charsReloaded[i]->GetDrawParent();
        if (dpOrig)
        {
            REQUIRE(dpReloaded != nullptr);
            CHECK(dpOrig->GetNameUTF8() == dpReloaded->GetNameUTF8());
        }
        else
        {
            CHECK(dpReloaded == nullptr);
        }
    }
}

TEST_CASE("MSB: round-trip preserves collision fields")
{
    auto& original = GetOriginalMSB();
    const auto writePath = GetOutDir() / "m10_00_00_00_roundtrip.msb";
    original.WriteToFilePath(writePath);
    auto reloaded = MSB::NewFromFilePath(writePath);
    REQUIRE(reloaded != nullptr);

    auto colsOrig     = original.GetPartParam().GetSubtypeEntries<CollisionPart>();
    auto colsReloaded = reloaded->GetPartParam().GetSubtypeEntries<CollisionPart>();
    REQUIRE(colsOrig.size() == colsReloaded.size());

    for (size_t i = 0; i < colsOrig.size(); ++i)
    {
        CHECK(colsOrig[i]->GetHitFilterId()  == colsReloaded[i]->GetHitFilterId());
        CHECK(colsOrig[i]->GetPlayRegionId() == colsReloaded[i]->GetPlayRegionId());
        CHECK(colsOrig[i]->GetEntityId()     == colsReloaded[i]->GetEntityId());
    }
}

TEST_CASE("MSB: round-trip preserves asset fields")
{
    auto& original = GetOriginalMSB();
    const auto writePath = GetOutDir() / "m10_00_00_00_roundtrip.msb";
    original.WriteToFilePath(writePath);
    auto reloaded = MSB::NewFromFilePath(writePath);
    REQUIRE(reloaded != nullptr);

    auto assetsOrig     = original.GetPartParam().GetSubtypeEntries<AssetPart>();
    auto assetsReloaded = reloaded->GetPartParam().GetSubtypeEntries<AssetPart>();
    REQUIRE(assetsOrig.size() == assetsReloaded.size());

    for (size_t i = 0; i < assetsOrig.size(); ++i)
    {
        CHECK(assetsOrig[i]->GetEntityId()           == assetsReloaded[i]->GetEntityId());
        CHECK(assetsOrig[i]->GetSfxParamRelativeId() == assetsReloaded[i]->GetSfxParamRelativeId());
    }
}

TEST_CASE("MSB: round-trip preserves region fields")
{
    auto& original = GetOriginalMSB();
    const auto writePath = GetOutDir() / "m10_00_00_00_roundtrip.msb";
    original.WriteToFilePath(writePath);
    auto reloaded = MSB::NewFromFilePath(writePath);
    REQUIRE(reloaded != nullptr);

    auto regionsOrig     = original.GetRegionParam().GetAllEntries();
    auto regionsReloaded = reloaded->GetRegionParam().GetAllEntries();
    REQUIRE(regionsOrig.size() == regionsReloaded.size());

    for (size_t i = 0; i < regionsOrig.size(); ++i)
    {
        CHECK(regionsOrig[i]->GetEntityId()  == regionsReloaded[i]->GetEntityId());
        CHECK(regionsOrig[i]->GetType()      == regionsReloaded[i]->GetType());
        CHECK(regionsOrig[i]->GetShapeType() == regionsReloaded[i]->GetShapeType());
    }
}

TEST_CASE("MSB: round-trip preserves event fields")
{
    auto& original = GetOriginalMSB();
    const auto writePath = GetOutDir() / "m10_00_00_00_roundtrip.msb";
    original.WriteToFilePath(writePath);
    auto reloaded = MSB::NewFromFilePath(writePath);
    REQUIRE(reloaded != nullptr);

    auto eventsOrig     = original.GetEventParam().GetAllEntries();
    auto eventsReloaded = reloaded->GetEventParam().GetAllEntries();
    REQUIRE(eventsOrig.size() == eventsReloaded.size());

    for (size_t i = 0; i < eventsOrig.size(); ++i)
    {
        CHECK(eventsOrig[i]->GetEntityId() == eventsReloaded[i]->GetEntityId());
        CHECK(eventsOrig[i]->GetType()     == eventsReloaded[i]->GetType());
    }
}

// ============================================================================
// Stable serialization (double write)
// ============================================================================

TEST_CASE("MSB: double write produces byte-identical files")
{
    auto& original = GetOriginalMSB();
    const auto path1 = GetOutDir() / "stable_write_1.msb";
    const auto path2 = GetOutDir() / "stable_write_2.msb";

    // First write.
    original.WriteToFilePath(path1);
    REQUIRE(exists(path1));

    // Re-read and second write.
    auto reloaded = MSB::NewFromFilePath(path1);
    REQUIRE(reloaded != nullptr);
    reloaded->WriteToFilePath(path2);
    REQUIRE(exists(path2));

    // Byte-for-byte comparison.
    auto bytes1 = ReadWholeFile(path1);
    auto bytes2 = ReadWholeFile(path2);
    REQUIRE(bytes1.size() == bytes2.size());

    bool identical = (memcmp(bytes1.data(), bytes2.data(), bytes1.size()) == 0);
    if (!identical)
    {
        for (size_t i = 0; i < bytes1.size(); ++i)
        {
            if (bytes1[i] != bytes2[i])
            {
                FAIL_CHECK("First diff at byte offset 0x", std::hex, i,
                           ": 0x", static_cast<int>(bytes1[i]),
                           " vs 0x", static_cast<int>(bytes2[i]));
                break;
            }
        }
    }
    CHECK(identical);
}

File: FirelinkER/tests/FirelinkERMaps/test_er_msb.py
"""Pytest suite for Elden Ring MSB via pyrelink_er_maps — mirrors test_er_msb.cpp.

Uses the bundled Stormveil Castle MSB (m10_00_00_00.msb, DCX removed) to verify:
  1. MSB loads without error.
  2. All param counts are positive.
  3. Known subtype distributions are non-empty.
  4. Names, types, and cross-references resolve correctly.
  5. repr() works on entries.
  6. Draw-parent cross-references can be mutated.
  7. Write -> re-read produces identical param counts and entry data.
  8. Double write produces byte-identical files (stable serialization).
"""

from __future__ import annotations

import math
from pathlib import Path

import pytest

from _pyrelink_er_maps import *

# ============================================================================
# Fixtures
# ============================================================================

RESOURCES_DIR = Path(__file__).resolve().parent.parent / "resources"
MSB_PATH = RESOURCES_DIR / "m10_00_00_00.msb"


@pytest.fixture(scope="session")
def original_msb() -> MSB:
    """Load the bundled MSB once for the whole test session."""
    assert MSB_PATH.exists(), f"MSB resource not found: {MSB_PATH}"
    return MSB.from_path(MSB_PATH)


@pytest.fixture()
def tmp_dir(tmp_path: Path) -> Path:
    return tmp_path


# ============================================================================
# Basic loading
# ============================================================================


def test_loads_without_error(original_msb: MSB):
    assert original_msb is not None


# ============================================================================
# Param counts
# ============================================================================


def test_all_param_counts_positive(original_msb: MSB):
    assert len(original_msb.model_param) > 0
    assert len(original_msb.event_param) > 0
    assert len(original_msb.region_param) > 0
    assert len(original_msb.route_param) > 0
    assert len(original_msb.part_param) > 0


# ============================================================================
# Subtype distributions
# ============================================================================


def test_model_subtype_distribution(original_msb: MSB):
    counts: dict[ModelType, int] = {}
    for m in original_msb.models:
        counts[m.model_type] = counts.get(m.model_type, 0) + 1
    assert counts.get(ModelType.MapPiece, 0) > 0
    assert counts.get(ModelType.Character, 0) > 0
    assert counts.get(ModelType.Collision, 0) > 0
    assert counts.get(ModelType.Asset, 0) > 0


def test_part_subtype_distribution(original_msb: MSB):
    counts: dict[PartType, int] = {}
    for p in original_msb.parts:
        counts[p.part_type] = counts.get(p.part_type, 0) + 1
    assert counts.get(PartType.MapPiece, 0) > 0
    assert counts.get(PartType.Character, 0) > 0
    assert counts.get(PartType.Collision, 0) > 0
    assert counts.get(PartType.Asset, 0) > 0


def test_region_count(original_msb: MSB):
    assert len(original_msb.region_param) > 10


def test_at_least_one_treasure_event(original_msb: MSB):
    treasures = [e for e in original_msb.events if e.event_type == EventType.Treasure]
    assert len(treasures) > 0


# ============================================================================
# Entry names and cross-references
# ============================================================================


def test_all_entry_names_nonempty(original_msb: MSB):
    for m in original_msb.models:
        assert m.name
    for e in original_msb.events:
        assert e.name
    for r in original_msb.regions:
        assert r.name
    for p in original_msb.parts:
        assert p.name


def test_every_part_has_model(original_msb: MSB):
    for p in original_msb.parts:
        assert p.model is not None, f"Part '{p.name}' has no model"


def test_first_part_model_in_model_list(original_msb: MSB):
    parts = original_msb.parts
    models = original_msb.models
    assert len(parts) > 0
    assert parts[0].model is not None
    model_name = parts[0].model.name
    assert any(m.name == model_name for m in models)


# ============================================================================
# repr
# ============================================================================


def test_entry_repr_nonempty(original_msb: MSB):
    parts = original_msb.parts
    assert len(parts) > 0
    assert repr(parts[0])

    characters = [p for p in parts if isinstance(p, CharacterPart)]
    assert len(characters) > 0
    assert repr(characters[0])

    if characters[0].model is not None:
        assert repr(characters[0].model)


# ============================================================================
# Character part fields
# ============================================================================


def test_character_translate_finite(original_msb: MSB):
    characters = [p for p in original_msb.parts if isinstance(p, CharacterPart)]
    assert len(characters) > 0
    x, y, z = characters[0].translate
    assert math.isfinite(x) and math.isfinite(y) and math.isfinite(z)


def test_character_scale_positive(original_msb: MSB):
    characters = [p for p in original_msb.parts if isinstance(p, CharacterPart)]
    assert len(characters) > 0
    sx, sy, sz = characters[0].scale
    assert sx > 0 and sy > 0 and sz > 0


# ============================================================================
# Collision part fields
# ============================================================================


def test_collision_hit_filter(original_msb: MSB):
    collisions = [p for p in original_msb.parts if isinstance(p, CollisionPart)]
    assert len(collisions) > 0
    hf = collisions[0].hit_filter_id
    assert 8 <= hf <= 29


# ============================================================================
# Region shapes
# ============================================================================


def test_all_regions_have_shape(original_msb: MSB):
    for r in original_msb.regions:
        assert r.shape is not None, f"Region '{r.name}' has no shape"


# ============================================================================
# Draw-parent mutation
# ============================================================================


def test_draw_parent_clear():
    msb = MSB.from_path(MSB_PATH)
    characters = [p for p in msb.parts if isinstance(p, CharacterPart)]
    assert len(characters) > 0
    characters[0].draw_parent = None
    assert characters[0].draw_parent is None


def test_draw_parent_reassign():
    msb = MSB.from_path(MSB_PATH)
    parts = msb.parts
    characters = [p for p in parts if isinstance(p, CharacterPart)]
    assert len(parts) > 0 and len(characters) > 0
    characters[0].draw_parent = parts[0]
    assert characters[0].draw_parent is not None
    assert characters[0].draw_parent.name == parts[0].name


# ============================================================================
# Round-trip: write -> re-read -> compare
# ============================================================================


def _roundtrip(original: MSB, tmp_dir: Path) -> MSB:
    path = tmp_dir / "roundtrip.msb"
    original.write_to_path(path)
    return MSB.from_path(path)


def test_roundtrip_param_counts(original_msb: MSB, tmp_dir: Path):
    reloaded = _roundtrip(original_msb, tmp_dir)
    assert len(original_msb.model_param) == len(reloaded.model_param)
    assert len(original_msb.event_param) == len(reloaded.event_param)
    assert len(original_msb.region_param) == len(reloaded.region_param)
    assert len(original_msb.route_param) == len(reloaded.route_param)
    assert len(original_msb.part_param) == len(reloaded.part_param)


def test_roundtrip_part_names(original_msb: MSB, tmp_dir: Path):
    reloaded = _roundtrip(original_msb, tmp_dir)
    orig_parts = original_msb.parts
    new_parts = reloaded.parts
    assert len(orig_parts) == len(new_parts)
    for a, b in zip(orig_parts, new_parts):
        assert a.name == b.name


def test_roundtrip_model_names(original_msb: MSB, tmp_dir: Path):
    reloaded = _roundtrip(original_msb, tmp_dir)
    orig = original_msb.models
    new = reloaded.models
    assert len(orig) == len(new)
    for a, b in zip(orig, new):
        assert a.name == b.name


def test_roundtrip_event_names(original_msb: MSB, tmp_dir: Path):
    reloaded = _roundtrip(original_msb, tmp_dir)
    orig = original_msb.events
    new = reloaded.events
    assert len(orig) == len(new)
    for a, b in zip(orig, new):
        assert a.name == b.name


def test_roundtrip_region_names(original_msb: MSB, tmp_dir: Path):
    reloaded = _roundtrip(original_msb, tmp_dir)
    orig = original_msb.regions
    new = reloaded.regions
    assert len(orig) == len(new)
    for a, b in zip(orig, new):
        assert a.name == b.name


def test_roundtrip_character_fields(original_msb: MSB, tmp_dir: Path):
    reloaded = _roundtrip(original_msb, tmp_dir)
    orig_chars = [p for p in original_msb.parts if isinstance(p, CharacterPart)]
    new_chars = [p for p in reloaded.parts if isinstance(p, CharacterPart)]
    assert len(orig_chars) == len(new_chars)

    for a, b in zip(orig_chars, new_chars):
        assert a.ai_id == b.ai_id
        assert a.character_id == b.character_id
        assert a.talk_id == b.talk_id
        assert a.entity_id == b.entity_id
        assert a.translate == b.translate

        dp_a = a.draw_parent
        dp_b = b.draw_parent
        if dp_a is not None:
            assert dp_b is not None
            assert dp_a.name == dp_b.name
        else:
            assert dp_b is None


def test_roundtrip_collision_fields(original_msb: MSB, tmp_dir: Path):
    reloaded = _roundtrip(original_msb, tmp_dir)
    orig = [p for p in original_msb.parts if isinstance(p, CollisionPart)]
    new = [p for p in reloaded.parts if isinstance(p, CollisionPart)]
    assert len(orig) == len(new)

    for a, b in zip(orig, new):
        assert a.hit_filter_id == b.hit_filter_id
        assert a.play_region_id == b.play_region_id
        assert a.entity_id == b.entity_id


def test_roundtrip_asset_fields(original_msb: MSB, tmp_dir: Path):
    reloaded = _roundtrip(original_msb, tmp_dir)
    orig = [p for p in original_msb.parts if isinstance(p, AssetPart)]
    new = [p for p in reloaded.parts if isinstance(p, AssetPart)]
    assert len(orig) == len(new)

    for a, b in zip(orig, new):
        assert a.entity_id == b.entity_id
        assert a.sfx_param_relative_id == b.sfx_param_relative_id


def test_roundtrip_region_fields(original_msb: MSB, tmp_dir: Path):
    reloaded = _roundtrip(original_msb, tmp_dir)
    orig = original_msb.regions
    new = reloaded.regions
    assert len(orig) == len(new)

    for a, b in zip(orig, new):
        assert a.entity_id == b.entity_id
        assert a.region_type == b.region_type
        assert a.shape_type == b.shape_type


def test_roundtrip_event_fields(original_msb: MSB, tmp_dir: Path):
    reloaded = _roundtrip(original_msb, tmp_dir)
    orig = original_msb.events
    new = reloaded.events
    assert len(orig) == len(new)

    for a, b in zip(orig, new):
        assert a.entity_id == b.entity_id
        assert a.event_type == b.event_type


# ============================================================================
# Stable serialization (double write)
# ============================================================================


def test_double_write_byte_identical(original_msb: MSB, tmp_dir: Path):
    path1 = tmp_dir / "stable_1.msb"
    path2 = tmp_dir / "stable_2.msb"

    original_msb.write_to_path(path1)
    assert path1.exists()

    reloaded = MSB.from_path(path1)
    reloaded.write_to_path(path2)
    assert path2.exists()

    bytes1 = path1.read_bytes()
    bytes2 = path2.read_bytes()
    assert len(bytes1) == len(bytes2)
    assert bytes1 == bytes2, (
        f"First diff at offset {next(i for i, (a, b) in enumerate(zip(bytes1, bytes2)) if a != b):#x}"
    )

